bouncer/src/repo/lib/repo_license.cpp

```diff
@@ -0,0 +1,169 @@
+		cryptolensHandle->machine_code_computer.set_machine_code(e, instanceId.toString());
+
+		// License activation 
+		cryptolens::optional<cryptolens::LicenseKey> licenseKey =
+			cryptolensHandle->activate_floating
+			(
+				e, // Object used for reporting if an error occured
+				authToken, // Cryptolens Access Token
+				productId, // Product key associated with the 3drepo.io pipeline
+				licenseStr, // License Key
+				floatingTimeIntervalSec // The amount of time the user has to wait before an actived machine (or session in our case) is taken off the license.
+			);
+
+		// dealing with early bail out scenarios
+		repoInfo << "****License activation summary****";
+		if (e)
+		{
+			cryptolens::ActivateError error = cryptolens::ActivateError::from_reason(e.get_reason());
+			repoInfo << "- server message: " << error.what();
+			repoInfo << "- server respose ok: false";
+			repoInfo << "- session not added to license";
+			throw repo::lib::RepoInvalidLicenseException();
+		}
+		else if (!licenseKey)
+		{
+			repoInfo << "- server respose ok: false";
+			repoInfo << "- session not added to license. Error license LicenseKey is null";
+			throw repo::lib::RepoInvalidLicenseException();
+		}
+		// dealing with the license check logic once all info is present
+		else
+		{
+			// printing out the result
+			std::string notes = licenseKey->get_notes().has_value() ?
+				licenseKey->get_notes().value() : "";
+			int noUsedInsances = licenseKey->get_activated_machines().has_value() ?
+				licenseKey->get_activated_machines()->size() : -1;
+			int maxInstances = licenseKey->get_maxnoofmachines().has_value() ?
+				licenseKey->get_maxnoofmachines().value() : -1;
+			bool licenseBlocked = licenseKey->get_block();
+			bool licenseExpired = licenseKey->check().has_expired(time(0)) ? true : false;
+			repoInfo << "- session license ID: " << instanceId.toString();
+			repoInfo << "- server message: " << notes;
+			repoInfo << "- server respose ok: true";
+			repoInfo << "- license blocked: " << licenseBlocked;
+			repoInfo << "- license expired: " << licenseExpired;
+			repoInfo << "- license expiry on: " <<
+				GetFormattedUtcTime(licenseKey->get_expires()) << " (UTC)";
+			if (noUsedInsances >= 0) repoInfo << "- activated instances: " << noUsedInsances;
+			if (maxInstances > 0) repoInfo << "- allowed instances: " << maxInstances;
+
+			// handle result
+			bool allCheck = !licenseExpired && !licenseBlocked;
+			if (allCheck)
+			{
+				repoInfo << "- activation result: session succesfully added to license";
+			}
+			else
+			{
+				repoInfo << "- activation result: session activation failed";
+				throw repo::lib::RepoInvalidLicenseException();
+			}
+		}
+#endif
+	}
+
+	void LicenseValidator::RunDeactivation()
+	{
+#ifdef REPO_LICENSE_CHECK
+		// only deactivate if we have succesfully activated
+		if (instanceId.isDefaultValue() && cryptolensHandle) return;
+
+		cryptolens::Error e;
+		cryptolensHandle->deactivate(
```

Comment:
do we need to store `instanceId`? it doesn't look like we use it here? I guess `cryptolensHandle` already contain all the necessary information?

Reply:
Also I think we want to deconstruct `cryptolensHandle` here and reset the variable.

Whilst it's fair to say that the current use case, the program is very likely to exit immediately, but it's probably more future proof to do it with the assumption that the user could potentially trigger the process again, instead of leaving the class in a weird state.

Reply:
I assume you mean that once `RunDeactivation` triggers, that we should clean up the `instanceId` and `cryptolensHandle`, so that another `RunActivation` call can be run succesfully in the same process? That's a good catch, will clean that up. I will carry across this change to the .NET implementation aswell.